### DISCUSSION TITLE
unexpected end of JSON input

### DISCUSSION THREAD

**Discussion by ghubz** [prometheus/prometheus#11091] [Q&A]
Hi,

I have a job reading targets from a json file. That json file is being updated via a cronjob.

The issue is that Prometheus complains with "unexpected end of JSON input" even though the json file is completely valid.

I tried different update time as well as different scrape intervals but no luck. What's weird is that I have other jobs with identical setup and they are OK. So I assume something in the JSON file, even though valid, causes prometheus to complain? Anyone had a similar issue?

**Comment by roidelapluie** [comment]
Are you writing the file atomically? (create a `.tmp` file and then `mv` it?)

**Comment by ghubz** [comment]
No, I am generating the content and then write it directly to the file (echo $content > list.json)

**Comment by ghubz** [comment]
Could it be a race condition? Once the file starts being updated Prometheus detects it and starts reading it however by the time it reads it the update hasn't finished so it always sees it as invalid.

**Comment by ghubz** [comment]
This is still ongoing and it is causing us major issues as we have multiple prometheus servers reporting the same error.

caller=file.go:337 level=error component="discovery manager scrape" discovery=file msg="Error reading file" path=/root/prometheus/targets.json err="unexpected end of JSON input"

The targets.json file is in **valid** JSON format and the file is just 800kb written at once from buffer via a shell script.

...any help appreciated...

**Comment by a-shoemaker** [comment]
I was looking for this same thing, this answers it: #12384 

It can happen due to the method not being an atomic write, so it's reading it while it's still being written.  E.g., so you can write it to a temporary file and then move it into place to address.
